core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,268 @@
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from refreshIf.test()", e);
+        }
+
+        return refresh;
+    }
+
+    private void load(@Nullable T cache, CompletableFuture<CacheEntry<T>> future) {
+        try {
+            requireNonNull(loader.apply(cache), "loader.apply() returned null")
+                    .handle((val, cause) -> {
+                        if (cause != null) {
+                            logger.warn("Failed to load a new value from loader: {}. the previous value: {}",
+                                        loader, cache, cause);
+                            handleException(cause, cache, future);
+                        } else {
+                            future.complete(new CacheEntry<>(val));
+                        }
+                        return null;
+                    });
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from loader.apply()", e);
+            handleException(e, cache, future);
+        }
+    }
+
+    private void handleException(Throwable cause, @Nullable T cache,
+                                 CompletableFuture<CacheEntry<T>> future) {
+        if (exceptionHandler != null) {
+            handleByExceptionHandler(cause, cache, future);
+        } else {
+            future.completeExceptionally(cause);
+        }
+    }
+
+    private void handleByExceptionHandler(Throwable originCause, @Nullable T cache,
+                                          CompletableFuture<CacheEntry<T>> future) {
+        try {
+            final CompletableFuture<T> fallback = exceptionHandler.apply(originCause, cache);
+            if (fallback != null) {
+                fallback.handle((val, cause) -> {
+                    if (cause != null) {
+                        logger.warn("Failed to load a new value from exceptionHandler: {}." +
+                                    "the previous value: {}", exceptionHandler, cache, cause);
+                        future.completeExceptionally(cause);
+                    } else {
+                        future.complete(new CacheEntry<>(val));
+                    }
+                    return null;
+                });
+            } else {
+                future.completeExceptionally(originCause);
+            }
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from exceptionHandler.apply()", e);
+            future.completeExceptionally(originCause);
+        }
+    }
+
+    private boolean isValid(@Nullable CacheEntry<T> cacheEntry) {
+        if (cacheEntry == null) {
+            return false;
+        }
+
+        if (expireAfterLoad != null) {
+            final long elapsed = System.nanoTime() - cacheEntry.cachedAt;
+            if (elapsed >= expireAfterLoad.toNanos()) {
+                logger.debug("CacheEntry {} is expired due to expireAfterLoad {}",
+                        cacheEntry, expireAfterLoad);
```

Comment:
It would be better to inform what was the cached value rather than directly exposing `CacheEntry` because `CacheEntry` is an internal term that users may not understand. Additionally `.toString()` is not implemented. 
```suggestion
                logger.debug("The cached value expired after {} ms. cache: {}",
                             expireAfterLoad.toMillis(), cacheEntry.value);
```

Reply:
✅ updated. nice suggestion!